### ISSUE TITLE
Better interface for "passing down keyword arguments"

### ISSUE THREAD

**Issue by gerritholl** [pytroll/satpy#1414]
## Feature Request

**Is your feature request related to a problem? Please describe.**

In multiple places in satpy customisation requires the user to pass nested dictionaries of keyword arguments.  For example, to save an animation using `MultiScene.save_animation`:

```python
ms.save_animation(..., enh_args={"decorate": {
    "decorate": [
        {"text": {
            "txt": "{start_time:%Y-%m-%d %H:%M}",
            "align": {
                "top_bottom": "bottom",
                "left_right": "right"},
            "font": '/usr/share/fonts/truetype/arial.ttf',
            "font_size": 20,
            "height": 30,
            "bg": "black",
            "bg_opacity": 255,
            "line": "white"}}]},
    "overlay": {
        "coast_dir": "/media/nas/x21308/shp/",
        "overlays": {
            "coasts": {
                "outline": (255, 255, 0),
                "width": 1.5,
                "level": 1,
                "resolution": "f"},
            "rivers": {
                "outline": (0, 0, 255),
                "width": 1.5,
                "level": 5,
                "resolution": "f"},
            "borders": {
                "outline": (0, 0, 0),
                "width": 1.5,
                "level": 3,
                "resolution": "f"}}}})
```

or for a simple image:

```python
ls.save_dataset(
        "natural_color",
        "/media/nas/x21308/plots_and_maps/2020/06/19/20200102T0915-andalus-natural-color.jpg",
        overlay={"coast_dir": "/media/nas/x21308/shp/",
                 "overlays": {
                     "points": {
                         "font": '/usr/share/fonts/truetype/arial.ttf',
                         "font_size": 30,
                         "points_list": [((-3.366, 36.927), "Mirador de Poqeira")],
                         "symbol": "circle",
                         "ptsize": 3,
                         "outline": "black",
                         "fill": "white",
                         "width": 20,
                         "fill_opacity": 255,
                         "box_outline": "black",
                         "box_linewidth": 2.0,
                         "box_fill": (200, 200, 200),
                         "box_opacity": 127},
                     "coasts": {
                         "outline": (255, 255, 0),
                         "width": 1.5,
                         "level": 1,
                         "resolution": "f"}}},
        fill_value=0)
```

which is at least 3 layers deep a dictionary for the image and 4 layers deep for the animation.  Also when passing `reader_kwargs` to a `Scene` or `MultiScene` on creation (see #1397 and #1400).  This type of interface is hard to document, hard to use, and error prone.

**Describe the solution you'd like**

I don't know.  Something else.  Some type of configuration class perhaps.  Needs coordination with other packages such as pyresample, pycoast, and pydecorate.

How is this solved in other libraries?  Can Matplotlib be an inspiration?

**Describe any changes to existing user workflow**

Unless we keep supporting the nested dictionaries, it will break everybody's code.

**Additional context**

Nested dictionaries work, but they're problematic.

**Comment by djhoese** [comment]
So I've put these thoughts in a couple other places and they apply to the "save_datasets" methods, but I'm not sure how to improve the reader kwargs too much. I think the main improvement that can be made to `save_dataset/s` is to separate the enhancing, overlay, and decorate into their own Scene methods. It requires more steps for the user (especially because enhancements are a pretty common need for most users), but allows for cleaner interfaces. The major complication here is in the few cases where readers don't want enhancements (ex. ninjo or cf writers). This requires the user to know whether or not enhancements are needed.

I see this being implemented and made backwards compatible with a new `._enhanced` property on the Scene object so `save_datasets` can know whether or not it needs to do enhancing (but also print out a deprecation warning, maybe?). This property could also be used by the overlay and decorate methods to say that enhancing has to be done first (at least in most cases I think it does).

**Comment by gerritholl** [comment]
Would [`dataclasses`](https://docs.python.org/3/library/dataclasses.html) be useful here?

**Comment by djhoese** [comment]
Are you suggesting a dataclass of arguments that get passed to lower-level components (writers, enhancements, etc)?

**Comment by gerritholl** [comment]
I'm not sure, maybe?  Just as a way to have a dedicated type.  Maybe not specifically a dataclass although it appears to be a convenient way of defining one (I've not used one yet).  Matplotlib has its [artist API](https://matplotlib.org/3.3.3/api/artist_api.html) which may be a source of inspiration as well. Both decorations and overlays could be classes deriving from a common artist-like base class (maybe even actually from the `Artist` class?  Or maybe not).

**Comment by djhoese** [comment]
I'm not sure how dataclasses would help here. Either the overlay/decorations are a separate interface on the Scene. If they are separate objects that get used, then I think the interfaces are simple enough that there doesn't need to be a method on the Scene to interact with them (compared to the complexity of an Artist). Like:

```
scn = Scene(...)
enhancer = Enhancer(...)
overlayer = Overlayer(...)
decorator = Decorator(...)
scn2 = scn.apply_image_modifier(enhancer)
scn3 = scn2.apply_image_modifier(overlayer)
scn4 = scn3.apply_image_modifier(decorator)
```

Versus:

```
scn2 = enhancer.change_scene(scn)
```

Versus:

```
scn2 = scn.enhance(...)
scn3 = scn2.add_overlays(...)
scn4 = scn3.add_decoration(...)
```

Side note: There is an interesting benefit to this in general where users can add one decoration or one overlay at a time instead of having to put a ton of stuff in one big call.

**Comment by pnuu** [comment]
Note that all of these operations should be per composite/dataset, not the whole scene. Otherwise the enhancement/overlay/decoration classes/methods will need to have knowledge of each individual composite/dataset that have been loaded and would become quite complex to define.

**Comment by djhoese** [comment]
Good point @pnuu. At least I think I understand part of what you're saying. Why would these classes/methods need knowledge of the DataArrays beyond the number of bands they have? The `add_overlays`-like methods could always take a set of names/DataIDs that they should be applied to. Otherwise in most cases I saw this as refactoring where the user provides their arguments (enhancement versus overlay versus decorate).

Regardless, you bring up an important concept. We might as well define per-DataArray interfaces for these as a satpy accessor (ex. `my_data_arr.satpy.enhance(...)`). There is also the idea of config-defined operations like enhancements versus user-defined operations like calling `.stretch` (overlays and decorations could be done from configs too, but aren't currently).

**Comment by pnuu** [comment]
> Why would these classes/methods need knowledge of the DataArrays beyond the number of bands they have?

Different composites have different enhancements, they might need different colours for the overlays/decorations, might have physical units that are wanted as colorbar, and so on. All these need to be configured, and if all the configurations are in the same object that are applied to the whole scene, they need to be defined per `DataID` (or `DataQuery`? Haven't yet really looked at the current implementation). The same naturally applies if they are done by the user to each individual composite/dataset, but then each of the operations are more explicit. But much more verbose, so :man_shrugging:

**Comment by djhoese** [comment]
Ok. I suppose that would be a limitation that I adopted from the current `save_dataset/s` methods. Currently you have to define your overlay/decorations for all datasets being saved. In the above example I suppose this would be handled by making multiple Scenes. So maybe you enhance using the enhancement configuration to make one new Scene. Then you take that enhanced Scene and make two new Scenes (with two separate calls): one Scene for all grayscale images and another Scene for the colorized ir composite (for example). Then you either join the Scenes back together before saving or you save each separately with two calls to `Scene.save_datasets()`.

**Comment by pnuu** [comment]
The enhancements would need to per composite, not per group of grayscale or RGB datesets. Every single composite has different set of enhancements, unless someone is using dynamic stretches (which won't work for RGBs). Having the composite and enhancement as a single unit would simplify this a lot. Like it was in the `mpop' era, but external in a config file, not within a Python file.

**Comment by djhoese** [comment]
Yes, like we have now. A YAML file? That's what I meant by "using the enhancement configuration", using the existing YAML configuration "decision tree" approach that is currently implemented.

**Comment by pnuu** [comment]
Oh, indeed. I was thinking more on the lines that the enhancements etc. above were a completely new system. I think I hadn't read the couple posts between the first and the `Enhancer()` one.

Personally I'd just put everything in the YAML files, so that each composite had _all_ their settings in a single place. But I guess there are use cases where a programmatical(?) interface is needed so that composites/enhancements/overlays could be done without config files, and then these separate methods could be handy. And under the hood they could be used with the config file(s), too.
